### PR TITLE
PoS Block Producer: TxnConnectStatusByIndex

### PR DESCRIPTION
**Pull request by AeonSw4n**
**TODO:**

During Blockchain's ProcessBlockPoS, validate that:
- `MsgDeSoBlock.TxnConnectStatusByIndex.Size() == len(MsgDeSoBlock.Txns)`
- `HashBitset(MsgDeSoBlock.TxnConnectStatusByIndex).IsEqual( MsgDeSoBlock.Header.TxnConnectStatusByIndexHash)`

### CODE REVIEW

File: lib/network.go [hunk 1]
```diff
@@ -2132,6 +2141,12 @@ func (msg *MsgDeSoHeader) EncodeHeaderVersion2(preSignature bool) ([]byte, error
 	// The Nonce and ExtraNonce fields are unused in version 2. We skip them
 	// during both encoding and decoding.
 
+	// TxnConnectStatusByIndexHash
+	if msg.TxnConnectStatusByIndexHash == nil {
```
Comment (lazynina) [inline lib/network.go:2145]: Does this need to be gated on some migration?

File: lib/network.go [hunk 2]
```diff
@@ -2346,6 +2361,13 @@ func DecodeHeaderVersion2(rr io.Reader) (*MsgDeSoHeader, error) {
 	retHeader.Nonce = 0
 	retHeader.ExtraNonce = 0
 
+	// TxnConnectStatusByIndexHash
+	retHeader.TxnConnectStatusByIndexHash = &BlockHash{}
```
Comment (lazynina) [inline lib/network.go:2365]: do we need some sort of new migration or header version to handle this?

File: lib/network.go [hunk 3]
```diff
@@ -2721,6 +2760,14 @@ func (msg *MsgDeSoBlock) FromBytes(data []byte) error {
 		}
 	}
 
+	// Version 2 blocks have a TxnStatusConnectedIndex attached to them.
+	if ret.Header.Version == HeaderVersion2 {
```
Comment (lazynina) [inline lib/network.go:2764]: okay I see - this is where the header migration is handled

File: lib/network.go (hunk 1 above)
Comment (tholonious) [reply lib/network.go:2145]: No migration needed here. The existence of the filed is gated by the header version value = 2

File: lib/network.go (hunk 2 above)
Comment (tholonious) [reply lib/network.go:2365]: No migration needed here. The header version above defines whether the field exists or not

File: lib/network_test.go [hunk 4]
```diff
@@ -189,38 +196,38 @@ func TestHeaderConversionAndReadWriteMessage(t *testing.T) {
...
-		assert.NoError(err)
+		require.NoError(err)
 
 		testHdr := NewMessage(MsgTypeHeader)
 		err = testHdr.FromBytes(data)
-		assert.NoError(err)
+		require.NoError(err)
 
-		assert.Equal(expectedBlockHeader, testHdr)
+		require.True(assert.Equal(expectedBlockHeader, testHdr))
```
Comment (tholonious) [inline lib/network_test.go:205]: Does this work?
```suggestion
		require.Equal(expectedBlockHeader, testHdr)
```

File: lib/network_test.go [hunk 5]
```diff
@@ -189,38 +196,38 @@ func TestHeaderConversionAndReadWriteMessage(t *testing.T) {
 	// versions we want to test.
 	for _, expectedBlockHeader := range expectedBlockHeadersToTest {
 		data, err := expectedBlockHeader.ToBytes(false)
-		assert.NoError(err)
+		require.NoError(err)
```
Comment (tholonious) [inline lib/network_test.go:199]: Why not pass in the testing instance `t` to the NoError function? Same comment for all of uses of require below
```suggestion
		require.NoError(t, err)
```

**Review by tholonious — CHANGES_REQUESTED** [review]
Code LGTM over all. The unit tests failed in BuildKite though: https://buildkite.com/deso/core/builds/3507#018aae2a-e1b3-470c-876b-0940ba91191e

Can you take a look at the failure?

Not sure why the BK status occasionally doesn't show up in Github. I'll look into it separately.

**Review by tholonious — APPROVED** [review]
LGTM! Feel free to merge this and follow up on the TODOs with a separate PR. The bitset size and Bitset hash equivalence check are nice utilities, but don't preclude this code form shipping. 🚀

**Pull request merged**